File: frontend/src/types/depth.ts
export interface ImageQualityReport {
  resolution: [number, number];
  aspect_ratio: number;
  sharpness_score: number;
  sharpness_label: "Excellent" | "Good" | "Fair" | "Blurry";
  brightness_mean: number;
  brightness_label: "Optimal" | "Overexposed" | "Underexposed";
  contrast_score: number;
  contrast_label: "High" | "Optimal" | "Low";
  saturation_mean: number;
  cloud_shadow_risk: "Low" | "Moderate" | "High";
  view_geometry: string;
  nadir_warning: boolean;
  suitability_score: number;
  suitability_label: "High" | "Medium" | "Low";
}

export interface DepthConfidenceReport {
  overall_confidence: "HIGH" | "MEDIUM" | "LOW";
  confidence_score: number;
  local_gradient_consistency: number;
  edge_alignment_score: number;
  flat_depth_detected: boolean;
  invalid_pixel_pct: number;
  reasons: string[];
}

export interface CalibrationConfig {
  mode: "reference" | "metadata" | "terrain" | "relative";
  known_object_name?: string | null;
  known_object_height_m?: number | null;
  ref_pixel_coords?: number[] | null;
  gsd_cm_per_pixel?: number | null;
  flight_altitude_m?: number | null;
  focal_length_mm?: number | null;
  sensor_width_mm?: number | null;
  scale_factor_m_per_unit?: number | null;
  ground_baseline_z?: number | null;
  is_calibrated: boolean;
  calibration_description?: string | null;
}

export interface BuildingMeasurement {
  id: string;
  name: string;
  bbox: [number, number, number, number]; // [x, y, width, height]
  polygon?: [number, number][] | null;
  camera_to_building_distance_m?: number | null;
  depth_p10_m?: number | null;
  depth_p50_m?: number | null;
  depth_p90_m?: number | null;
  depth_min_m?: number | null;
  depth_max_m?: number | null;
  relative_depth?: number;
  rooftop_peak_z_rel: number;
  ground_base_z_rel: number;
  relative_height_unitless: number;
  calibrated_height_m?: number | null;
  ground_elevation_m?: number | null;
  roof_elevation_m?: number | null;
  height_fused_m?: number | null;
  height_depth_cue_m?: number | null;
  height_geom_cue_m?: number | null;
  height_shadow_cue_m?: number | null;
  height_ref_cue_m?: number | null;
  ground_truth_height_m?: number | null;
  error_m?: number | null;
  relative_error_pct?: number | null;
  accuracy_pct?: number | null;
  uncertainty_margin_m?: number | null;
  confidence_pct: number;
  is_metric_available: boolean;
  measurement_mode: string;
  calibration_note: string;
  scale_recovery_method?: string;
  ground_plane_method?: string;
  metric_status?: string;
  stage_status?: string;
  calculation_status?: string;
  geometry_diagnostics?: Record<string, any>;
  camera_parameter_status?: Record<string, string>;
  technical_derivation?: Record<string, any>;
}

export interface PointCloudData {
  points_count: number;
  positions: number[]; // flattened [x, y, z, ...]
  colors: number[];    // flattened [r, g, b, ...]
  bounds: {
    min_x: number;
    max_x: number;
    min_y: number;
    max_y: number;
    min_z: number;
    max_z: number;
  };
}

export interface MeshHeightfieldData {
  grid_width: number;
  grid_height: number;
  heights: number[];
  colors: number[];
}

export interface DepthProcessResponse {
  id: string;
  timestamp: string;
  scene_name: string;
  image_type: string;
  depth_backend: string;
  depth_backend_is_neural: boolean;
  image_url: string;
  depth_map_url: string;
  depth_colormap_turbo_url: string;
  depth_colormap_viridis_url: string;
  depth_colormap_magma_url: string;

  raw_depth_stats: {
    min: number;
    max: number;
    mean: number;
    std?: number;
    shape: [number, number];
  };
  metric_depth_stats?: {
    min_m: number;
    max_m: number;
    mean_m: number;
    median_m: number;
    valid_pixel_pct: number;
    unit: string;
  };
  camera_intrinsics?: {
    fx: number;
    fy: number;
    cx: number;
    cy: number;
    status: string;
  };
  depth_disclaimer?: string;
  quality_report: ImageQualityReport;
  confidence_report: DepthConfidenceReport;
  pointcloud: PointCloudData;
  mesh: MeshHeightfieldData;
  calibration: CalibrationConfig;
  sample_buildings: BuildingMeasurement[];
  timing_ms: {
    preprocessing_ms: number;
    depth_inference_ms: number;
    geometry_reconstruction_ms: number;
    total_ms: number;
  };
  scientific_disclaimer: string;
}

export interface ProcessImageRequest {
  image_url?: string | null;
  image_base64?: string | null;
  image_type: "satellite" | "aerial" | "drone";
  scene_preset_id?: string | null;
  calibration?: CalibrationConfig | null;
  camera_intrinsics?: Record<string, number> | null;
  model_variant?: "metric_base" | "metric_small" | "relative_small" | "structural" | null;
}

export interface MeasureBuildingRequest {
  scene_id?: string | null;
  bbox: [number, number, number, number];
  calibration?: CalibrationConfig | null;
  depth_matrix_id?: string | null;
  name?: string | null;
  ground_truth_height_m?: number | null;
}

export interface ValidationMeasurement {
  building_id: string;
  building_name: string;
  gt_height_m: number;
  pred_height_m: number;
  error_m: number;
  rel_error_pct: number;
  confidence_pct: number;
}

export interface ValidationBenchmark {
  scene_id: string;
  scene_name: string;
  sensor_type: string;
  lidar_source: string;
  buildings_tested: number;
  mae_m: number;
  rmse_m: number;
  median_err_m: number;
  mean_relative_error_pct: number;
  pct_within_2m?: number;
  pct_within_5m?: number;
  pct_within_10m?: number;
  sample_measurements: ValidationMeasurement[];
}

export interface ValidationOverview {
  data_available: boolean;
  unavailable_reason?: string | null;
  requirements: string[];
  total_scenes_evaluated: number;
  total_structures_measured: number;
  overall_mae_m?: number | null;
  overall_rmse_m?: number | null;
  overall_median_err_m?: number | null;
  overall_mre_pct?: number | null;
  overall_pct_within_2m?: number | null;
  overall_pct_within_5m?: number | null;
  overall_pct_within_10m?: number | null;
  benchmarks: ValidationBenchmark[];
  scatter_points: {
    scene: string;
    scene_id: string;
    building: string;
    gt_height: number;
    pred_height: number;
    error: number;
    sensor: string;
  }[];
}


export interface HeroSceneSummary {
  id: string;
  title: string;
  type: string;
  description: string;
  gt_benchmark_mae: string;
  image_url: string;
  view_angle: string;
  sensor: string;
}

export interface ValidationPoint {
  x: number;
  y: number;
  measured_depth_m: number;
  label?: string | null;
}

export interface ModelBenchmarkResult {
  model_key: string;
  model_name: string;
  status: "SUCCESS" | "UNAVAILABLE" | "INFERENCE_FAILED";
  device: string;
  device_info: {
    device?: string;
    param_count?: string;
    precision?: string;
    [key: string]: any;
  };
  inference_time_ms: number;
  preprocessing_time_ms: number;
  postprocessing_time_ms: number;
  total_time_ms: number;
  depth_min_m?: number | null;
  depth_max_m?: number | null;
  depth_mean_m?: number | null;
  depth_median_m?: number | null;
  depth_p10_m?: number | null;
  depth_p25_m?: number | null;
  depth_p50_m?: number | null;
  depth_p75_m?: number | null;
  depth_p90_m?: number | null;
  depth_std_m?: number | null;
  depth_spread_m?: number | null;
  p10_p90_spread_m?: number | null;
  valid_pixel_ratio?: number | null;
  is_compressed?: boolean;
  compression_note?: string | null;
  error_reason?: string | null;
  depth_colormap_url?: string | null;
  building_stats?: {
    building_name: string;
    bbox: number[];
    depth_p10_m: number;
    depth_p50_m: number;
    depth_p90_m: number;
    mean_depth_m: number;
    std_depth_m: number;
    stage3_predicted_height_m?: number | null;
    stage3_ground_elevation_m?: number | null;
    stage3_roof_elevation_m?: number | null;
    stage3_status?: string;
  }[];
  stage1_validation_metrics?: Record<string, any> | null;
  stage3_stress_test?: {
    building_name: string;
    metric_depth_p50_m: number;
    predicted_height_m?: number | null;
    ground_truth_height_m: number;
    absolute_error_m?: number | null;
    relative_error_pct?: number | null;
    accuracy_pct?: number | null;
    stage3_status?: string;
    vertical_reference?: string;
  } | null;
}

export interface BenchmarkRequest {
  scene_id?: string | null;
  image_url?: string | null;
  image_base64?: string | null;
  image_type?: string;
  selected_models: string[];
  validation_points?: ValidationPoint[];
  probe_pixel?: [number, number] | null;
  selected_bbox?: [number, number, number, number] | null;
}

export interface BenchmarkResponse {
  timestamp: string;
  scene_id?: string | null;
  image_dimensions: [number, number];
  selected_models: string[];
  model_results: ModelBenchmarkResult[];
  model_agreement: {
    agreement_rating: string;
    mean_p50_m?: number | null;
    std_p50_m?: number | null;
    spread_p50_m?: number | null;
    relative_std_pct?: number | null;
    models_compared_count?: number;
    note?: string;
  };
  pixel_probe?: {
    x: number;
    y: number;
    model_predictions: {
      model_key: string;
      model_name: string;
      depth_m: number;
    }[];
  } | null;
  region_analysis?: {
    bbox: number[];
    model_predictions: {
      model_key: string;
      model_name: string;
      min_m: number;
      max_m: number;
      mean_m: number;
      median_m: number;
      p10_m: number;
      p25_m: number;
      p50_m: number;
      p75_m: number;
      p90_m: number;
      std_m: number;
    }[];
  } | null;
  validation_summary?: {
    model_key: string;
    model_name: string;
    valid_samples_count: number;
    mae_m?: number | null;
    rmse_m?: number | null;
    mean_relative_error_pct?: number | null;
    median_relative_error_pct?: number | null;
    point_evaluations: {
      x: number;
      y: number;
      label: string;
      predicted_depth_m: number;
      measured_depth_m: number;
      abs_error_m: number;
      rel_error_pct: number;
    }[];
  }[];
  best_model_recommendation: {
    best_model_key?: string | null;
    best_model_name: string;
    basis?: string;
    recommendation: string;
    next_step: string;
    mae_m?: number;
    rmse_m?: number;
  };
}

export interface QualityGateReport {
  is_passed: boolean;
  status_label: string;
  dataset_sufficiency: "LOW" | "MEDIUM" | "GOOD" | "NONE";
  sufficiency_reason: string;
  total_samples: number;
  total_scenes: number;
  train_samples: number;
  val_samples: number;
  test_split_samples: number;
  detected_depth_units: string[];
  depth_min_observed_m?: number | null;
  depth_max_observed_m?: number | null;
  valid_pixel_ratio_mean?: number | null;
  leakage_detected: boolean;
  warnings: string[];
  blockers: string[];
  required_dataset_schema: Record<string, any>;
}

export interface TrainingState {
  status: string;
  mode?: string;
  is_running: boolean;
  current_epoch: number;
  total_epochs: number;
  progress_pct: number;
  last_run_summary?: Record<string, any> | null;
  error_message?: string | null;
}

export interface ModelRangeAnalysis {
  model_name: string;
  model_max_depth_m: number;
  dataset_max_depth_m: number;
  compatibility_status: string;
  is_compatible: boolean;
}

export interface TrainingStatusResponse {
  training_state: TrainingState;
  quality_gate_report: QualityGateReport;
  model_range_analysis: ModelRangeAnalysis;
  checkpoint_available: boolean;
  checkpoint_path?: string | null;
  training_summary?: Record<string, any> | null;
  required_dataset_specifications: Record<string, any>;
}

export interface GAMUSDatasetReport {
  status: "VALID" | "INVALID" | "CORRUPTED" | "EMPTY";
  data_dir: string;
  sample_counts: {
    total: number;
    train: number;
    val: number;
    test: number;
  };
  rgb_dimensions: [number, number];
  semantic_classes: Record<string, number>;
  agl_stats: {
    min?: number;
    max?: number;
    mean?: number;
    median?: number;
    valid_ratio?: number;
  };
  warnings: string[];
  errors: string[];
  corrupted_files: string[];
}

export interface TrainingModeInfo {
  status: "AVAILABLE" | "BLOCKED";
  requirements: string[];
  description: string;
  blocked_reason?: string | null;
}

export interface MultiTaskTrainingStatusResponse {
  training_state: TrainingState;
  datasets: {
    gamus: {
      status: "AVAILABLE" | "MISSING";
      details: GAMUSDatasetReport;
    };
    metric_depth: {
      status: "AVAILABLE" | "MISSING";
      details: QualityGateReport;
    };
  };
  modes_availability: {
    ZERO_SHOT: TrainingModeInfo;
    AERIAL_DOMAIN_ADAPTATION: TrainingModeInfo;
    AERIAL_METRIC_FINE_TUNING: TrainingModeInfo;
  };
  last_report?: Record<string, any> | null;
}

File: backend/models/depth.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
import uuid
from datetime import datetime

class ImageQualityReport(BaseModel):
    resolution: List[int] = Field(description="[Width, Height] in pixels")
    aspect_ratio: float
    sharpness_score: float = Field(description="Laplacian variance score")
    sharpness_label: str = Field(description="Excellent, Good, Fair, or Blurry")
    brightness_mean: float = Field(description="Mean grayscale brightness 0-255")
    brightness_label: str = Field(description="Optimal, Overexposed, or Underexposed")
    contrast_score: float = Field(description="RMS contrast")
    contrast_label: str = Field(description="High, Optimal, or Low")
    saturation_mean: float = Field(description="Mean HSV saturation 0-1")
    cloud_shadow_risk: str = Field(description="Low, Moderate, or High")
    view_geometry: str = Field(description="Oblique (25-45°), Near-Nadir (75-90°), or Low-Altitude Drone")
    nadir_warning: bool = Field(description="True if near-nadir angle with limited vertical facade cues")
    image_quality_score: float = Field(default=80.0, description="Visual image quality score 0-100")
    depth_quality_score: float = Field(default=80.0, description="Monocular depth quality score 0-100")
    geometry_quality_score: float = Field(default=75.0, description="Geometric structure & view angle quality 0-100")
    metric_calibration_quality_score: float = Field(default=20.0, description="Metric calibration readiness score 0-100")
    height_suitability_score: float = Field(default=30.0, description="Suitability score specifically for vertical height estimation 0-100")
    height_suitability_label: str = Field(default="Low", description="High, Medium, or Low")
    height_suitability_reason: str = Field(default="Low observability for metric height — oblique imagery recommended.", description="Detailed explanation of height observability")
    suitability_score: float = Field(description="Overall depth suitability 0-100")
    suitability_label: str = Field(description="High, Medium, or Low")


class DepthConfidenceReport(BaseModel):
    overall_confidence: str = Field(description="HIGH, MEDIUM, or LOW")
    confidence_score: float = Field(description="0-100 quality confidence score")
    local_gradient_consistency: float
    edge_alignment_score: float
    flat_depth_detected: bool
    invalid_pixel_pct: float
    reasons: List[str] = Field(default_factory=list)

class CalibrationConfig(BaseModel):
    mode: str = Field(default="relative", description="reference, metadata, terrain, or relative")
    known_object_name: Optional[str] = None
    known_object_height_m: Optional[float] = None
    ref_pixel_coords: Optional[List[int]] = None  # [x, y, w, h]
    gsd_cm_per_pixel: Optional[float] = None
    flight_altitude_m: Optional[float] = None
    focal_length_mm: Optional[float] = None
    sensor_width_mm: Optional[float] = None
    scale_factor_m_per_unit: Optional[float] = 1.0
    ground_baseline_z: Optional[float] = 0.0
    is_calibrated: bool = False
    calibration_description: Optional[str] = "Relative depth — metric scale unavailable."

class BuildingMeasurement(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    name: str
    bbox: List[int] = Field(description="[x, y, width, height] in image coordinates")
    polygon: Optional[List[List[int]]] = None
    camera_to_building_distance_m: Optional[float] = Field(default=None, description="Estimated camera-to-building metric distance in metres")
    depth_p10_m: Optional[float] = Field(default=None, description="10th percentile building depth in metres")
    depth_p50_m: Optional[float] = Field(default=None, description="50th percentile (median) building depth in metres")
    depth_p90_m: Optional[float] = Field(default=None, description="90th percentile building depth in metres")
    depth_min_m: Optional[float] = Field(default=None, description="Minimum building depth in metres")
    depth_max_m: Optional[float] = Field(default=None, description="Maximum building depth in metres")
    relative_depth: float = Field(default=0.0, description="Raw relative depth value at building top")
    rooftop_peak_z_rel: float
    ground_base_z_rel: float
    relative_height_unitless: float
    calibrated_height_m: Optional[float] = Field(default=None, description="Building vertical height in metres (Stage 3 3D unprojection)")
    ground_elevation_m: Optional[float] = Field(default=None, description="Terrain base ground elevation in metres")
    roof_elevation_m: Optional[float] = Field(default=None, description="Building roof peak elevation in metres")
    height_fused_m: Optional[float] = None
    height_depth_cue_m: Optional[float] = None
    height_geom_cue_m: Optional[float] = None
    height_shadow_cue_m: Optional[float] = None
    height_ref_cue_m: Optional[float] = None
    ground_truth_height_m: Optional[float] = None
    error_m: Optional[float] = None
    relative_error_pct: Optional[float] = None
    accuracy_pct: Optional[float] = None
    uncertainty_margin_m: Optional[float] = None
    confidence_pct: float
    is_metric_available: bool
    measurement_mode: str
    calibration_note: str
    scale_recovery_method: Optional[str] = "Stage 3 Metric 3D Unprojection"
    ground_plane_method: Optional[str] = "RANSAC / Perimeter Ring Datum"
    metric_status: str = "Relative depth — metric scale unavailable."
    stage_status: str = "Stage 3 Under Geometric Validation"
    calculation_status: str = Field(default="UNAVAILABLE", description="UNAVAILABLE, GEOMETRY_READY, VERTICAL_REFERENCE_READY, HEIGHT_COMPUTED, or VALIDATION_AVAILABLE")
    geometry_diagnostics: Dict[str, Any] = Field(default_factory=dict)
    camera_parameter_status: Dict[str, str] = Field(
        default_factory=lambda: {
            "focal_length": "UNAVAILABLE",
            "sensor_width": "UNAVAILABLE",
            "altitude": "UNAVAILABLE",
            "view_angle": "ESTIMATED"
        }
    )
    technical_derivation: Dict[str, Any] = Field(default_factory=dict)

class HeightEstimationRequest(BaseModel):
    scene_id: Optional[str] = None
    bbox: List[int] = Field(description="[x, y, width, height]")
    camera_pose: Optional[Dict[str, float]] = Field(default=None, description="pitch_deg, roll_deg, yaw_deg")
    ground_reference_bbox: Optional[List[int]] = None
    roof_reference_bbox: Optional[List[int]] = None
    use_ground_reference: bool = False
    use_roof_reference: bool = False

class HeightEstimationResponse(BaseModel):
    height_available: bool
    estimated_height_m: Optional[float] = None
    ground_elevation_m: Optional[float] = None
    roof_elevation_m: Optional[float] = None
    height_uncertainty_m: Optional[float] = None
    confidence: float
    confidence_pct: float
    method: str
    camera_pose_available: bool
    camera_pose_status: str = "Unavailable"
    vertical_reference: str = "Estimated Ground-Plane Normal"
    intrinsics_source: str
    ground_plane_inlier_ratio: float
    roof_plane_inlier_ratio: float
    stage_status: str = "Stage 3 Under Geometric Validation"
    calculation_status: str = Field(default="UNAVAILABLE", description="UNAVAILABLE, GEOMETRY_READY, VERTICAL_REFERENCE_READY, HEIGHT_COMPUTED, or VALIDATION_AVAILABLE")
    reason_if_unavailable: Optional[str] = None
    geometry_diagnostics: Dict[str, Any] = Field(default_factory=dict)
    technical_derivation: Dict[str, Any] = Field(default_factory=dict)

class HeightValidationRequest(BaseModel):
    estimated_height_m: Optional[float] = Field(default=None, description="Estimated building vertical height in metres")
    ground_truth_height_m: float = Field(description="Independent surveyed ground truth height in metres")

class HeightValidationResponse(BaseModel):
    ground_truth_height_m: float
    estimated_height_m: Optional[float] = None
    absolute_error_m: Optional[float] = None
    relative_error_percent: Optional[float] = None
    accuracy_percent: Optional[float] = None
    evaluation_note: str

class PointCloudData(BaseModel):
    points_count: int
    positions: List[float] = Field(description="Flattened camera-space [x, y, z, ...]")
    colors: List[float] = Field(description="Flattened [r, g, b, ...]")
    bounds: Dict[str, float]

class MeshHeightfieldData(BaseModel):
    grid_width: int
    grid_height: int
    heights: List[float] = Field(description="Flattened 2D grid height values")
    colors: List[float] = Field(description="Flattened 2D grid RGB color values")

class ProcessImageRequest(BaseModel):
    image_url: Optional[str] = None
    image_base64: Optional[str] = None
    image_type: str = Field(default="aerial", description="satellite, aerial, or drone")
    scene_preset_id: Optional[str] = None
    calibration: Optional[CalibrationConfig] = None
    camera_intrinsics: Optional[Dict[str, float]] = None
    model_variant: Optional[str] = Field(default="metric_base", description="metric_base, metric_small, relative_small, or structural")

class DepthProcessResponse(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    scene_name: str
    image_type: str
    depth_backend: str = Field(description="Name of the depth engine that actually produced this result")
    depth_backend_is_neural: bool = Field(description="True only if real learned weights performed the inference")
    image_url: str
    depth_map_url: str
    depth_colormap_turbo_url: str
    depth_colormap_viridis_url: str
    depth_colormap_magma_url: str
    raw_depth_stats: Dict[str, Any]
    metric_depth_stats: Dict[str, Any] = Field(default_factory=dict, description="Min, Max, Mean, Median depth in metres")
    camera_intrinsics: Dict[str, Any] = Field(default_factory=dict, description="fx, fy, cx, cy, status")
    depth_disclaimer: str = Field(default="Metric depth produced by pretrained outdoor metric-depth model.")
    quality_report: ImageQualityReport
    confidence_report: DepthConfidenceReport
    pointcloud: PointCloudData
    mesh: MeshHeightfieldData
    calibration: CalibrationConfig
    sample_buildings: List[BuildingMeasurement]
    timing_ms: Dict[str, float]
    scientific_disclaimer: str

class MeasureBuildingRequest(BaseModel):
    scene_id: Optional[str] = None
    bbox: List[int] = Field(description="[x, y, width, height]")
    calibration: Optional[CalibrationConfig] = None
    depth_matrix_id: Optional[str] = None
    name: Optional[str] = "Target Structure"
    ground_truth_height_m: Optional[float] = Field(default=None, description="Optional surveyed ground-truth height for error scoring")

class ValidationBenchmark(BaseModel):
    scene_id: str
    scene_name: str
    sensor_type: str
    lidar_source: str
    buildings_tested: int
    mae_m: float
    rmse_m: float
    median_err_m: float
    mean_relative_error_pct: float
    pct_within_2m: float = 0.0
    pct_within_5m: float = 0.0
    pct_within_10m: float = 0.0
    sample_measurements: List[Dict[str, Any]]

class ValidationOverview(BaseModel):
    data_available: bool = Field(
        description="False when no verified reference dataset is registered; metrics are then meaningless"
    )
    unavailable_reason: Optional[str] = Field(
        default=None, description="Why validation cannot be computed"
    )
    requirements: List[str] = Field(
        default_factory=list, description="What is required to populate real validation"
    )
    total_scenes_evaluated: int
    total_structures_measured: int
    overall_mae_m: Optional[float] = None
    overall_rmse_m: Optional[float] = None
    overall_median_err_m: Optional[float] = None
    overall_mre_pct: Optional[float] = None
    overall_pct_within_2m: Optional[float] = None
    overall_pct_within_5m: Optional[float] = None
    overall_pct_within_10m: Optional[float] = None
    benchmarks: List[ValidationBenchmark]
    scatter_points: List[Dict[str, Any]]

class HeroSceneSummary(BaseModel):
    id: str
    title: str
    type: str
    description: str
    gt_benchmark_mae: str
    image_url: str
    view_angle: str
    sensor: str


class ValidationPoint(BaseModel):
    x: int = Field(description="Pixel X coordinate")
    y: int = Field(description="Pixel Y coordinate")
    measured_depth_m: float = Field(description="Actual measured camera-to-surface distance in metres")
    label: Optional[str] = Field(default=None, description="Optional point label (e.g. Ground Control 1)")


class ModelBenchmarkResult(BaseModel):
    model_key: str
    model_name: str
    status: str = Field(description="SUCCESS, UNAVAILABLE, or INFERENCE_FAILED")
    device: str = "cpu"
    device_info: Dict[str, Any] = Field(default_factory=dict)
    inference_time_ms: float
    preprocessing_time_ms: float = 0.0
    postprocessing_time_ms: float = 0.0
    total_time_ms: float = 0.0
    depth_min_m: Optional[float] = None
    depth_max_m: Optional[float] = None
    depth_mean_m: Optional[float] = None
    depth_median_m: Optional[float] = None
    depth_p10_m: Optional[float] = None
    depth_p25_m: Optional[float] = None
    depth_p50_m: Optional[float] = None
    depth_p75_m: Optional[float] = None
    depth_p90_m: Optional[float] = None
    depth_std_m: Optional[float] = None
    depth_spread_m: Optional[float] = None
    p10_p90_spread_m: Optional[float] = None
    valid_pixel_ratio: Optional[float] = None
    is_compressed: bool = False
    compression_note: Optional[str] = None
    error_reason: Optional[str] = None
    depth_colormap_url: Optional[str] = None
    building_stats: List[Dict[str, Any]] = Field(default_factory=list)
    stage1_validation_metrics: Optional[Dict[str, Any]] = None
    stage3_stress_test: Optional[Dict[str, Any]] = Field(default=None, description="FROZEN Stage 3 predicted height + Stage 4 validation against 175m")


class BenchmarkRequest(BaseModel):
    scene_id: Optional[str] = None
    image_url: Optional[str] = None
    image_base64: Optional[str] = None
    image_type: str = Field(default="aerial", description="aerial, drone, or satellite")
    selected_models: List[str] = Field(
        default=["da_v2_base", "da_v2_large", "unidepth_v2_l", "metric3dv2"],
        description="Keys of models to benchmark"
    )
    validation_points: Optional[List[ValidationPoint]] = Field(default_factory=list)
    probe_pixel: Optional[List[int]] = Field(default=None, description="[x, y] pixel coordinates to probe across models")
    selected_bbox: Optional[List[int]] = Field(default=None, description="[x, y, w, h] region to analyze across models")


class BenchmarkResponse(BaseModel):
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    scene_id: Optional[str] = None
    image_dimensions: List[int]
    selected_models: List[str]
    model_results: List[ModelBenchmarkResult]
    model_agreement: Dict[str, Any] = Field(default_factory=dict)
    pixel_probe: Optional[Dict[str, Any]] = None
    region_analysis: Optional[Dict[str, Any]] = None
    validation_summary: Optional[List[Dict[str, Any]]] = Field(default_factory=list)
    best_model_recommendation: Dict[str, Any] = Field(default_factory=dict)

